artemis/modules/gau.py

```diff
@@ -0,0 +1,62 @@
+#!/usr/bin/env python3
+import urllib.parse
+
+from karton.core import Task
+
+from artemis.binds import TaskStatus, TaskType
+from artemis.config import Config
+from artemis.domains import is_subdomain
+from artemis.module_base import ArtemisBase
+from artemis.utils import check_output_log_on_error
+
+
+class GAU(ArtemisBase):
+    """
+    Consumes `type: domain` and adds subdomains fetched from AlienVault's Open Threat Exchange, the Wayback Machine,
+    and Common Crawl using https://github.com/lc/gau. Produces `type: domain`.
+    """
+
+    identity = "gau"
+    filters = [
+        {"type": TaskType.DOMAIN.value},
+    ]
+    # We don't ensure we're the only module touching the target, as actually we interact with the providers, not
+    # the target. We lock the providers instead in the run() method.
+    lock_target = False
+
+    def run(self, current_task: Task) -> None:
+        domain = current_task.get_payload("domain")
+        with self.lock:
+            output = check_output_log_on_error(
+                [
+                    "gau",
+                ]
+                + Config.GAU_ADDITIONAL_OPTIONS,
+                self.log,
+                input=domain.encode("ascii"),
+            )
+
+            domains = set()
+            for line in output.decode("ascii", errors="ignore").split("\n"):
+                url_parsed = urllib.parse.urlparse(line)
+                if not url_parsed.hostname:
+                    continue
+
+                if is_subdomain(url_parsed.hostname, domain):
+                    domains.add(url_parsed.hostname)
+
+            for domain in domains:
+                task = Task(
+                    {"type": TaskType.DOMAIN},
```

Comment:
Not missing a `.value` here?

Reply:
On a related note, how do you mitigate feedback loops? This karton service is producing the same headers it accepts, which if I'm not mistaken, can lead to an endless loop

Reply:
I don't think .value is needed here, the karton I copied this from works without

regarding feedback loops, here are two kinds of loops possible:

- adding a record that already exists in a cyclical way (e.g. domain -> ip -> revdns -> domain) which is prevented by add_task() being a wrapper that makes sure such task doesn't already exist,
- adding an endless stream of different e.g. subdomains (domain -> git.subdomain -> git.git.subdomain ->  ...) which is not prevented but has also not been observed yet